src/Microsoft.Restier.Core/InvocationContextExtensions.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Restier.Core
+{
+    /// <summary>
+    /// Represents the API engine and provides a set of static
+    /// (Shared in Visual Basic) methods for interacting with objects
+    /// that implement <see cref="InvocationContext"/>.
+    /// </summary>
+    public static class InvocationContextExtensions
+    {
+        #region GetApiService<T>
+
+        /// <summary>
+        /// Gets an API service.
+        /// </summary>
+        /// <param name="context">
+        /// An invocation context.
+        /// </param>
+        /// <typeparam name="T">The API service type.</typeparam>
+        /// <returns>The API service instance.</returns>
+        public static T GetApiService<T>(this InvocationContext context) where T : class
+        {
+            Ensure.NotNull(context, "context");
+            return context.ServiceProvider.GetService<T>();
+        }
+
+        /// <summary>
+        /// Gets an ordered collection of service instances.
+        /// </summary>
+        /// <param name="context">
+        /// An invocation context.
+        /// </param>
+        /// <typeparam name="T">The API service type.</typeparam>
+        /// <returns>The ordered collection of service instances.</returns>
+        public static IEnumerable<T> GetApiServices<T>(this InvocationContext context) where T : class
+        {
+            Ensure.NotNull(context, "context");
+            return context.ServiceProvider.GetServices<T>();
+        }
+
+        /// <summary>
+        /// Gets a service from the <see cref="ApiContext"/>.
+        /// </summary>
+        /// <param name="context">
+        /// An invocation context.
+        /// </param>
+        /// <typeparam name="T">The service type.</typeparam>
+        /// <returns>The service instance.</returns>
+        public static T GetApiContextService<T>(this InvocationContext context) where T : class
+        {
+            Ensure.NotNull(context, "context");
+            return context.ApiContext.GetApiService<T>();
```

Comment:
I'm inclined not to get services from ApiContext, if we decide to give InvocationContext its own scope, we'd better assume every service resolving is targeting InvocationContext, instead of ApiContext.
We only have to assign ApiHolder.Api for the scope of InvocationContext. Or if it's possible to be more aggressive, we could remove the ApiContext scope, and make InvocationContext the primary parameter of all those service calls. If we want more compatibility, maybe inherit InvocationContext from ApiContext?